dd-sdk-android/src/main/cpp/utils/signal-monitor.c

```diff
@@ -0,0 +1,163 @@
+#include "signal-monitor.h"
+#include <errno.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <stdbool.h>
+#include <jni.h>
+#include "../datadog-native-lib.h"
+
+#define HANDLED_SIGNAL_COUNT 6
+
+void handleSignal(int signum, siginfo_t *info,
+                  void *user_context);
+
+bool configureSignalStack(void);
+
+stack_t signalStack;
+/**
+ * Global shared context for Bugsnag reports
+ */
+static JNIEnv *globalEnv;
+/* the signal handler array */
+struct sigaction *globalSigaction;
+
+/* the previous signal handler array */
+struct sigaction *globalSigactionPrevious;
+
+/**
+ * Native signals which will be captured by the Bugsnag signal handler™
+ */
+static const int native_signals[HANDLED_SIGNAL_COUNT + 1] = {
+        SIGILL, SIGTRAP, SIGABRT, SIGBUS, SIGFPE, SIGSEGV};
+static const char signal_name[HANDLED_SIGNAL_COUNT + 1][8] = {
+        "SIGILL", "SIGTRAP", "SIGABRT", "SIGBUS", "SIGFPE", "SIGSEGV"};
+static const char signal_generic_error_message[HANDLED_SIGNAL_COUNT + 1][60] = {
+        "Illegal instruction",
+        "Trace/breakpoint trap",
+        "Abort program",
+        "Bus error (bad memory access)",
+        "Floating-point exception",
+        "Segmentation violation (invalid memory reference)"};
+
+
+bool installSignalHandlers(JNIEnv *env) {
+    if (globalEnv != NULL) {
+        return true; // already installed
+    }
+    static pthread_mutex_t handlerMutex = PTHREAD_MUTEX_INITIALIZER;
+    pthread_mutex_lock(&handlerMutex);
+    if (!configureSignalStack()) {
+        pthread_mutex_unlock(&handlerMutex);
+        return false;
+    }
+
+    globalEnv = env;
+    globalSigaction =
+            calloc(sizeof(struct sigaction), HANDLED_SIGNAL_COUNT);
+    if (globalSigaction == NULL) {
+        pthread_mutex_unlock(&handlerMutex);
+        return false;
+    }
+    sigemptyset(&globalSigaction->sa_mask);
+    globalSigaction->sa_sigaction = handleSignal;
+    globalSigaction->sa_flags = SA_SIGINFO | SA_ONSTACK;
+
+    globalSigactionPrevious =
+            calloc(sizeof(struct sigaction), HANDLED_SIGNAL_COUNT);
+    if (globalSigactionPrevious == NULL) {
+        pthread_mutex_unlock(&handlerMutex);
+        return false;
+    }
+    for (int i = 0; i < HANDLED_SIGNAL_COUNT; i++) {
+        const int signal = native_signals[i];
+        int success = sigaction(signal, globalSigaction,
+                                &globalSigactionPrevious[i]);
+        if (success != 0) {
+            pthread_mutex_unlock(&handlerMutex);
+            return false;
+        }
+    }
+
+    pthread_mutex_unlock(&handlerMutex);
+
+    return true;
+}
+
+void uninstallSignalHandlers() {
+    if (globalEnv == NULL)
```

Comment:
No mutex again, smells fishy!

Reply:
Hmm...yeah...my bad that I did not check that code before adding it :)